File: leetcode/algorithms/705_design_hash_set/main.js
/**
 * Your MyHashSet object will be instantiated and called as such:
 * var obj = new MyHashSet()
 * obj.add(key)
 * obj.remove(key)
 * var param_3 = obj.contains(key)
 */

var MyHashSet = function () {
  this.nums = Array(1000001).fill(0);
};

/**
 * @param {number} key
 * @return {void}
 */
MyHashSet.prototype.add = function (key) {
  this.nums[key] = 1;
};

/**
 * @param {number} key
 * @return {void}
 */
MyHashSet.prototype.remove = function (key) {
  this.nums[key] = 0;
};

/**
 * @param {number} key
 * @return {boolean}
 */
MyHashSet.prototype.contains = function (key) {
  return this.nums[key] === 1;
};


// Solution
var Solution_MyHashSet = function () {
  // Really you should just
  // Make your own object, but instead
  // we have attached ourself to the
  // `this` object which then becomes our hashmap.
  // What you should instead do is this:
  // this.hash_map = {}
  // And then update our following functions
};

Solution_MyHashSet.prototype.add = function (key) {
  // Constant Time
  // Linear Space | To the size of the input key
  // You can access objects using array notation

  this[key] = null;
};

Solution_MyHashSet.prototype.remove = function (key) {
  // Constant Time
  // Constant Space
  // You can access objects using array notation
  // Here we use the delete keyword.

  delete this[key];
};

Solution_MyHashSet.prototype.contains = function (key) {
  // Constant Time
  // Constant Space
  // This just asks if the property exists

  return this.hasOwnProperty(key);
};

File: leetcode/algorithms/705_design_hash_set/main.java
/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet obj = new MyHashSet();
 * obj.add(key);
 * obj.remove(key);
 * boolean param_3 = obj.contains(key);
 */

class MyHashSet {
    private int[] nums;

    public MyHashSet() {
        this.nums = new int[1000001];
    }

    public void add(int key) {
        this.nums[key] = 1;
    }

    public void remove(int key) {
        this.nums[key] = 0;
    }

    public boolean contains(int key) {
        return this.nums[key] == 1;
    }
}


// Solution
class Solution_MyHashSet {
    int num[];

    public Solution_MyHashSet() {
        num = new int[31251];
    }

    // set the bit if that element is present
    public void add(int key) {
        num[getIdx(key)] |= getMask(key);
    }

    // unset the bit if a key is not present
    public void remove(int key) {
        num[getIdx(key)] &= (~getMask(key));
    }

    // check if bit corresponding to the key is set or not
    public boolean contains(int key) {
        return (num[getIdx(key)] & getMask(key)) != 0;
    }

    // idx of num[] to which this key belongs
    // for key = 37, it will give 1
    private int getIdx(int key) {
        return (key / 32);
    }

    // get mask representing the bit inside num[idx] that corresponds to given key.
    // for key = 37, it will give 00000000000000000000000000100000
    private int getMask(int key) {
        key %= 32;
        return (1 << key);
    }
}
